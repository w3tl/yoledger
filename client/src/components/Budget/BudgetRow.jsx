import React from 'react';
import PropTypes from 'prop-types';
import BudgetCell from './BudgetCell';
import { accountPropType, budgetPropType } from '../propTypes';

class BudgetRow extends React.Component {
  static propTypes = {
    account: accountPropType.isRequired,
    budgets: PropTypes.arrayOf(budgetPropType),
    onUpdate: PropTypes.func.isRequired,
    onDelete: PropTypes.func.isRequired,
  }

  handleUpdate = period => ({ amount }) => {
    const { onUpdate, account } = this.props;
    onUpdate({ account, period, amount });
  }

  handleDelete = period => () => {
    const { onDelete, account } = this.props;
    onDelete({ account, period });
  }

  render() {
    const { account, budgets } = this.props;

    return (
      <tr>
        <td>{account.name}</td>
        {budgets.map(budget => (
          <td key={`${account.name}_${budget.period.dateStart}`}>
            <BudgetCell
              amount={budget.amount}
              onUpdate={this.handleUpdate(budget.period)}
              onDelete={this.handleDelete(budget.period)}
            />
          </td>
        ))}
      </tr>
    );
  }
}

BudgetRow.defaultProps = {
  budgets: [],
};

export default BudgetRow;
