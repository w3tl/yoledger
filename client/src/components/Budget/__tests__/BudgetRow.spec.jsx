import React from 'react';
import { configure, shallow, mount } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';
import Adapter from 'enzyme-adapter-react-16';
import BudgetRow from '../BudgetRow';

configure({ adapter: new Adapter() });

const accounts = [{
  name: 'Train',
}];

const budgets = [{
  id: '1',
  amount: 10,
  period: {
    dateStart: '2018-06-10',
  },
}, {
  id: '2',
  amount: 15,
  period: {
    dateStart: '2018-06-25',
  },
}, {
  id: '3',
  amount: 13,
  period: {
    dateStart: '2018-07-10',
  },
}];

describe('BudgetRow', () => {
  it('should render correctly', () => {
    const output = shallow(
      <BudgetRow
        account={accounts[0]}
        budgets={budgets}
        onUpdate={() => {}}
        onDelete={() => {}}
      />,
    );
    expect(shallowToJson(output)).toMatchSnapshot();
  });

  it('should add an account and a period to the events', () => {
    const onUpdate = jest.fn();
    const onDelete = jest.fn();
    const output = mount(
      <table>
        <tbody>
          <BudgetRow
            account={accounts[0]}
            budgets={budgets}
            onUpdate={onUpdate}
            onDelete={onDelete}
          />
        </tbody>
      </table>,
    );

    output.find('Button').first().simulate('click');
    expect(onUpdate).toHaveBeenCalledWith({
      account: accounts[0],
      period: budgets[0].period,
      amount: 10,
    });
  });
});
