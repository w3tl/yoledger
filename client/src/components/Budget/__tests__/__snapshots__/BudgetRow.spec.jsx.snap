// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BudgetRow should render correctly 1`] = `
<tr>
  <td>
    Train
  </td>
  <td
    key="Train_2018-06-10"
  >
    <BudgetCell
      amount={10}
      onDelete={[Function]}
      onUpdate={[Function]}
    />
  </td>
  <td
    key="Train_2018-06-25"
  >
    <BudgetCell
      amount={15}
      onDelete={[Function]}
      onUpdate={[Function]}
    />
  </td>
  <td
    key="Train_2018-07-10"
  >
    <BudgetCell
      amount={13}
      onDelete={[Function]}
      onUpdate={[Function]}
    />
  </td>
</tr>
`;
